### PR TITLE
Implementation of finite automata in C

### PR DESCRIPTION
**Pull request by abdouskamel**
Hello everybody,

Here is an implementation of finite automata in C. It includes :

- Recognizing a string.
- Eliminating not accessible and not co-accessible states.
- Complete the DFA.
- Complement the DFA.

Hope you enjoy it.

### CODE REVIEW

File: code/string_algorithms/src/finite_automata/C/dfa.c [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+  if (dfa->states == NULL)
+    return (-1);
+
+  dfa->transitions = malloc(sizeof(state_id_t*) * states_nb);
+  if (dfa->transitions == NULL) {
+    free(dfa->states);
+    return (-1);
+  }
+
+  int i, j;
```
Comment (arnavb) [inline code/string_algorithms/src/finite_automata/C/dfa.c]: Move variables close to where they are used; `j` should be near its loop.

File: code/string_algorithms/src/finite_automata/C/types.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+/* States are labelled by positive integers instead of strings */
+typedef int state_id_t;
+
+/* A symbol is a single char */
+typedef char symbol_t;
+
+typedef enum dfs_color_e
+{
+  WHITE, BLACK
```
Comment (arnavb) [inline code/string_algorithms/src/finite_automata/C/types.h]: Maybe move each enum to its own line

File: code/string_algorithms/src/finite_automata/C/dfa.c (hunk 1 above)
Comment (abdouskamel) [reply code/string_algorithms/src/finite_automata/C/dfa.c]: If I do so, I'd declare `j` under the loop.

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**